docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,321 @@
+}
+
+type VsphereClusterAccountSpec struct {
+    // CredentialsRef references a Secret with the credentials to use
+    CredentialsRef *corev1.ObjectReference `json:"credentialsRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using an array of namespaces or with a label selector.
+    // If this object is nil, no namespaces will be allowed
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)
+    // If both NamespaceList and Selector are defined, the union of the two would be used to determine if a namespace is allowed to use the account
+    //
+    // Allowed namespaces listed out by name
+    // +optional
+    // +nullable
+    NamespaceList []string `json:"list"`
+
+    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VsphereClusterSpec struct {
+    ...
+    // +optional
+    IdentityRef *corev1.TypedLocalObjectReference `json:"IdentityRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+- If IdentityRef is specified, and it references a `VsphereClusterAccount`, the CRD is fetched and used to create a session
+- If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session
+- The controller will add the VsphereCluster to OwnerReferences on a Secret that contains the account credentials
+- If IdentityRef is not specified, the controller will fallback to using credentials provided in the controller deployment
+- The session will be cached using the existing logic where the key is server + username + datacenter
+- The `IdentityRef` on a `VsphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
+- The controller will fail to delete a `VsphereClusterAccount` with an error if the account is used by `VsphereClusters`
+
+#### Clusterctl Changes
+
+Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VsphereClusterAccount` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.
```

Comment:
```suggestion
Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VSphereClusterAccount` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.
```